File: tools/update_prayers.mjs
// Fetch the mosque's timetable and publish it as prayers.json.
//
// Hardened over the original: retries with backoff, request timeouts, schema
// validation, and a date check — because the failure that matters is not an
// error, it is silently publishing yesterday's times or an empty payload and
// leaving every device to act on it.
//
// Providers are pluggable so a new mosque can be added by writing a preset in
// mosques/, without touching this file.

import fs from 'node:fs';
import path from 'node:path';

const ROOT = path.resolve(path.dirname(new URL(import.meta.url).pathname), '..');
const OUT = path.join(ROOT, 'prayers.json');
const TIMEOUT_MS = 25_000;
const RETRIES = 4;

const readJson = (p, d = null) => { try { return JSON.parse(fs.readFileSync(p, 'utf8')); } catch { return d; } };

// Redact secrets before anything reaches the CI log.
const scrub = (s, secrets) => secrets.filter(Boolean).reduce((acc, v) => acc.split(v).join('***'), String(s));

function ymdIn(tz, offsetDays = 0) {
  const d = new Date(Date.now() + offsetDays * 86_400_000);
  const parts = new Intl.DateTimeFormat('en-CA', {
    timeZone: tz, year: 'numeric', month: '2-digit', day: '2-digit',
  }).formatToParts(d);
  const get = t => parts.find(p => p.type === t).value;
  return `${get('year')}-${get('month')}-${get('day')}`;
}

function hmsIn(tz) {
  return new Intl.DateTimeFormat('en-CA', {
    timeZone: tz, hour: '2-digit', minute: '2-digit', second: '2-digit', hour12: false,
  }).format(new Date());
}

async function getJson(url, headers, secrets) {
  let lastErr;
  for (let attempt = 1; attempt <= RETRIES; attempt++) {
    const ac = new AbortController();
    const timer = setTimeout(() => ac.abort(), TIMEOUT_MS);
    try {
      const res = await fetch(url, { signal: ac.signal, headers });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      return await res.json();
    } catch (err) {
      lastErr = err;
      console.warn(`  attempt ${attempt}/${RETRIES} failed: ${scrub(err.message, secrets)}`);
      if (attempt < RETRIES) await new Promise(r => setTimeout(r, attempt * 3000));
    } finally { clearTimeout(timer); }
  }
  throw new Error(`giving up: ${scrub(lastErr?.message ?? 'unknown', secrets)}`);
}

// --- providers -------------------------------------------------------------

async function fetchMasjidal({ tz }) {
  const apiKey = process.env.PRAYER_API_KEY;
  const baseUrl = process.env.PRAYER_API_BASE_URL;
  if (!apiKey || !baseUrl) throw new Error('PRAYER_API_KEY / PRAYER_API_BASE_URL are not set');
  const secrets = [apiKey, baseUrl];
  const url = `${baseUrl}&day=${ymdIn(tz)}&time=${hmsIn(tz)}`;
  return {
    payload: await getJson(url, {
      accept: '*/*',
      'addin-api-key': apiKey,
      'user-agent': 'prayer-times-sync/2',
    }, secrets),
    secrets,
  };
}

// Public fallback provider: no key, works for any coordinates on earth, so a
// new mosque can be added without an account anywhere.
async function fetchAladhan({ tz, preset }) {
  const { latitude, longitude } = preset.location ?? {};
  if (latitude == null || longitude == null) throw new Error('aladhan provider needs location.latitude/longitude');
  const methodMap = { MWL: 3, ISNA: 2, EGYPT: 5, MAKKAH: 4, KARACHI: 1, TEHRAN: 7, JAFARI: 0 };
  const method = methodMap[(preset.calculation?.method ?? 'ISNA').toUpperCase()] ?? 2;
  const school = (preset.calculation?.asr ?? 'standard').toLowerCase() === 'hanafi' ? 1 : 0;
  const [y, m, d] = ymdIn(tz).split('-');
  const url = `https://api.aladhan.com/v1/timings/${d}-${m}-${y}`
    + `?latitude=${latitude}&longitude=${longitude}&method=${method}&school=${school}`;
  const res = await getJson(url, { accept: 'application/json' }, []);
  const t = res?.data?.timings;
  if (!t) throw new Error('aladhan returned no timings');
  const hm = v => `${String(v).slice(0, 5)}:00`;
  // Reshape into the same envelope the devices already understand.
  return {
    payload: {
      data: {
        name: preset.name ?? preset.id,
        city: preset.city ?? '',
        prayers: null,
        prayerOfDay: {
          prayerDate: `${ymdIn(tz)}T00:00:00`,
          singlePrayers: [
            { prayerName: 'Fajr',    prayerBegins: hm(t.Fajr),    prayerAdhan: hm(t.Fajr),    prayerIqamah: null },
            { prayerName: 'Sunrise', prayerBegins: null,          prayerAdhan: hm(t.Sunrise), prayerIqamah: null },
            { prayerName: 'Dhuhr',   prayerBegins: hm(t.Dhuhr),   prayerAdhan: hm(t.Dhuhr),   prayerIqamah: null },
            { prayerName: 'Asr',     prayerBegins: hm(t.Asr),     prayerAdhan: hm(t.Asr),     prayerIqamah: null },
            { prayerName: 'Sunset',  prayerBegins: null,          prayerAdhan: hm(t.Sunset),  prayerIqamah: null },
            { prayerName: 'Maghrib', prayerBegins: hm(t.Maghrib), prayerAdhan: hm(t.Maghrib), prayerIqamah: null },
            { prayerName: 'Isha',    prayerBegins: hm(t.Isha),    prayerAdhan: hm(t.Isha),    prayerIqamah: null },
          ],
        },
      },
      status: 'ALADHAN',
      message: 'Success',
      source: 'aladhan',
    },
    secrets: [],
  };
}

// --- validation ------------------------------------------------------------

const REQUIRED = ['Fajr', 'Dhuhr', 'Asr', 'Maghrib', 'Isha'];

function validate(payload, tz) {
  const day = payload?.data?.prayerOfDay;
  if (!day) throw new Error('payload has no data.prayerOfDay');

  const list = day.singlePrayers;
  if (!Array.isArray(list) || list.length === 0) throw new Error('singlePrayers is empty');

  const date = String(day.prayerDate ?? '').slice(0, 10);
  const allowed = [ymdIn(tz, -1), ymdIn(tz), ymdIn(tz, 1)];
  if (!allowed.includes(date)) {
    throw new Error(`timetable is for ${date || '(none)'}, expected one of ${allowed.join(', ')}`);
  }
  if (date !== ymdIn(tz)) console.warn(`  note: payload date ${date} is not today (${ymdIn(tz)})`);

  const byName = new Map(list.map(p => [p.prayerName, p]));
  const missing = REQUIRED.filter(n => !byName.has(n));
  if (missing.length) throw new Error(`missing prayers: ${missing.join(', ')}`);

  const timeRe = /^\d{1,2}:\d{2}(:\d{2})?$/;
  for (const n of REQUIRED) {
    const p = byName.get(n);
    const anchors = [p.prayerAdhan, p.prayerBegins, p.prayerIqamah].filter(v => v && timeRe.test(v));
    if (anchors.length === 0) throw new Error(`${n} has no usable time (adhan/begins/iqamah all absent or malformed)`);
  }
  return { date, count: list.length };
}

// --- main ------------------------------------------------------------------

async function main() {
  const config = readJson(path.join(ROOT, 'config.json'), {}) ?? {};
  const mosqueId = process.env.PRAYER_MOSQUE || config.mosque || 'masjid-el-noor';
  const preset = readJson(path.join(ROOT, 'mosques', `${mosqueId}.json`), {}) ?? {};
  const tz = config.timezone ?? preset.timezone ?? 'UTC';
  const provider = preset.timetable?.provider ?? 'masjidal';

  console.log(`mosque=${mosqueId} tz=${tz} provider=${provider} today=${ymdIn(tz)}`);

  const providers = { masjidal: fetchMasjidal, aladhan: fetchAladhan };
  const fn = providers[provider];
  if (!fn) throw new Error(`unknown timetable provider '${provider}'`);

  const { payload, secrets } = await fn({ tz, preset });
  const info = validate(payload, tz);
  console.log(`validated: ${info.count} entries for ${info.date}`);

  const next = JSON.stringify(payload, null, 2) + '\n';
  const prev = fs.existsSync(OUT) ? fs.readFileSync(OUT, 'utf8') : '';
  if (prev === next) {
    console.log('unchanged');
    return;
  }
  // Write via a temp file so an interrupted run cannot leave a truncated
  // prayers.json for every device to download.
  const tmp = `${OUT}.tmp`;
  fs.writeFileSync(tmp, next);
  fs.renameSync(tmp, OUT);
  console.log(`wrote prayers.json for ${info.date}`);
  void scrub;
  void secrets;
}

main().catch(err => { console.error('fatal:', err.message); process.exit(1); });

File: tools/update_prayers.py
#!/usr/bin/env python3
"""Fetch the mosque's timetable and publish it as prayers.json.

Providers are pluggable so a new mosque can be added by writing a preset in
mosques/, without touching this file.

The failure that matters is not an error: it is silently publishing
yesterday's times or an empty payload and leaving every device to act on it.
So: retries with backoff, request timeouts, schema validation, and a date
check before anything is written.
"""

from __future__ import annotations

import json
import os
import sys
import time
import urllib.request
from datetime import datetime, timedelta
from pathlib import Path

try:
    from zoneinfo import ZoneInfo
except ImportError:  # pragma: no cover - Python < 3.9
    print("python 3.9+ required (for zoneinfo)", file=sys.stderr)
    sys.exit(2)

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "prayers.json"
TIMEOUT_S = 25
RETRIES = 4
UA = "prayer-times-sync/2"


def scrub(text, secrets):
    out = str(text)
    for s in secrets:
        if s:
            out = out.replace(s, "***")
    return out


def read_json(path: Path, default=None):
    try:
        with path.open(encoding="utf-8") as fh:
            return json.load(fh)
    except (OSError, json.JSONDecodeError):
        return default


def ymd_in(tz_name: str, offset_days: int = 0) -> str:
    d = datetime.now(ZoneInfo(tz_name)) + timedelta(days=offset_days)
    return d.strftime("%Y-%m-%d")


def hms_in(tz_name: str) -> str:
    return datetime.now(ZoneInfo(tz_name)).strftime("%H:%M:%S")


def get_json(url: str, headers: dict, secrets: list) -> dict:
    last_err = None
    for attempt in range(1, RETRIES + 1):
        try:
            req = urllib.request.Request(url, headers=headers)
            with urllib.request.urlopen(req, timeout=TIMEOUT_S) as resp:
                status = getattr(resp, "status", 200)
                if status < 200 or status >= 300:
                    raise RuntimeError(f"HTTP {status}")
                return json.loads(resp.read().decode("utf-8", errors="replace"))
        except Exception as exc:  # noqa: BLE001 - any failure is just a retry
            last_err = exc
            print(f"  attempt {attempt}/{RETRIES} failed: {scrub(exc, secrets)}", file=sys.stderr)
            if attempt < RETRIES:
                time.sleep(attempt * 3)
    raise RuntimeError(f"giving up: {scrub(last_err or 'unknown', secrets)}")


def fetch_masjidal(tz: str, preset: dict):
    api_key = os.environ.get("PRAYER_API_KEY", "")
    base_url = os.environ.get("PRAYER_API_BASE_URL", "")
    if not api_key or not base_url:
        raise RuntimeError("PRAYER_API_KEY / PRAYER_API_BASE_URL are not set")
    url = f"{base_url}&day={ymd_in(tz)}&time={hms_in(tz)}"
    secrets = [api_key, base_url]
    payload = get_json(url, {
        "accept": "*/*",
        "addin-api-key": api_key,
        "user-agent": UA,
    }, secrets)
    return payload, secrets


def fetch_aladhan(tz: str, preset: dict):
    loc = preset.get("location", {}) or {}
    lat = loc.get("latitude")
    lng = loc.get("longitude")
    if lat is None or lng is None:
        raise RuntimeError("aladhan provider needs location.latitude/longitude")
    calc = preset.get("calculation", {}) or {}
    method_map = {"MWL": 3, "ISNA": 2, "EGYPT": 5, "MAKKAH": 4,
                  "KARACHI": 1, "TEHRAN": 7, "JAFARI": 0}
    method = method_map.get(str(calc.get("method", "ISNA")).upper(), 2)
    school = 1 if str(calc.get("asr", "standard")).lower() == "hanafi" else 0
    y, m, d = ymd_in(tz).split("-")
    url = (f"https://api.aladhan.com/v1/timings/{d}-{m}-{y}"
           f"?latitude={lat}&longitude={lng}&method={method}&school={school}")
    res = get_json(url, {"accept": "application/json"}, [])
    t = ((res or {}).get("data") or {}).get("timings")
    if not t:
        raise RuntimeError("aladhan returned no timings")

    def hm(v):
        return f"{str(v)[:5]}:00"

    # Reshape into the envelope the devices already understand.
    payload = {
        "data": {
            "name": preset.get("name") or preset.get("id"),
            "city": preset.get("city", ""),
            "prayers": None,
            "prayerOfDay": {
                "prayerDate": f"{ymd_in(tz)}T00:00:00",
                "singlePrayers": [
                    {"prayerName": "Fajr",    "prayerBegins": hm(t["Fajr"]),    "prayerAdhan": hm(t["Fajr"]),    "prayerIqamah": None},
                    {"prayerName": "Sunrise", "prayerBegins": None,             "prayerAdhan": hm(t["Sunrise"]), "prayerIqamah": None},
                    {"prayerName": "Dhuhr",   "prayerBegins": hm(t["Dhuhr"]),   "prayerAdhan": hm(t["Dhuhr"]),   "prayerIqamah": None},
                    {"prayerName": "Asr",     "prayerBegins": hm(t["Asr"]),     "prayerAdhan": hm(t["Asr"]),     "prayerIqamah": None},
                    {"prayerName": "Sunset",  "prayerBegins": None,             "prayerAdhan": hm(t["Sunset"]),  "prayerIqamah": None},
                    {"prayerName": "Maghrib", "prayerBegins": hm(t["Maghrib"]), "prayerAdhan": hm(t["Maghrib"]), "prayerIqamah": None},
                    {"prayerName": "Isha",    "prayerBegins": hm(t["Isha"]),    "prayerAdhan": hm(t["Isha"]),    "prayerIqamah": None},
                ],
            },
        },
        "status": "ALADHAN",
        "message": "Success",
        "source": "aladhan",
    }
    return payload, []


REQUIRED = ("Fajr", "Dhuhr", "Asr", "Maghrib", "Isha")


def _valid_time(v) -> bool:
    if not isinstance(v, str):
        return False
    parts = v.split(":")
    if len(parts) not in (2, 3):
        return False
    return all(p.isdigit() for p in parts)


def validate(payload: dict, tz: str):
    day = ((payload or {}).get("data") or {}).get("prayerOfDay")
    if not day:
        raise RuntimeError("payload has no data.prayerOfDay")

    lst = day.get("singlePrayers")
    if not isinstance(lst, list) or not lst:
        raise RuntimeError("singlePrayers is empty")

    date_str = str(day.get("prayerDate") or "")[:10]
    allowed = [ymd_in(tz, -1), ymd_in(tz), ymd_in(tz, 1)]
    if date_str not in allowed:
        raise RuntimeError(
            f"timetable is for {date_str or '(none)'}, expected one of {', '.join(allowed)}"
        )
    if date_str != ymd_in(tz):
        print(f"  note: payload date {date_str} is not today ({ymd_in(tz)})", file=sys.stderr)

    by_name = {e.get("prayerName"): e for e in lst if isinstance(e, dict)}
    missing = [n for n in REQUIRED if n not in by_name]
    if missing:
        raise RuntimeError(f"missing prayers: {', '.join(missing)}")

    for n in REQUIRED:
        p = by_name[n]
        anchors = [p.get("prayerAdhan"), p.get("prayerBegins"), p.get("prayerIqamah")]
        if not any(_valid_time(a) for a in anchors):
            raise RuntimeError(
                f"{n} has no usable time (adhan/begins/iqamah all absent or malformed)"
            )

    return date_str, len(lst)


PROVIDERS = {"masjidal": fetch_masjidal, "aladhan": fetch_aladhan}


def main() -> int:
    config = read_json(ROOT / "config.json", {}) or {}
    mosque_id = os.environ.get("PRAYER_MOSQUE") or config.get("mosque") or "masjid-el-noor"
    preset = read_json(ROOT / "mosques" / f"{mosque_id}.json", {}) or {}
    tz = config.get("timezone") or preset.get("timezone") or "UTC"
    provider = (preset.get("timetable") or {}).get("provider", "masjidal")

    print(f"mosque={mosque_id} tz={tz} provider={provider} today={ymd_in(tz)}")

    fn = PROVIDERS.get(provider)
    if fn is None:
        raise RuntimeError(f"unknown timetable provider '{provider}'")

    payload, _secrets = fn(tz, preset)
    date_str, count = validate(payload, tz)
    print(f"validated: {count} entries for {date_str}")

    nxt = json.dumps(payload, indent=2) + "\n"
    prev = OUT.read_text(encoding="utf-8") if OUT.exists() else ""
    if prev == nxt:
        print("unchanged")
        return 0

    # Write via a temp file so an interrupted run cannot leave a truncated
    # prayers.json for every device to download.
    tmp = OUT.parent / f"{OUT.name}.tmp"
    tmp.write_text(nxt, encoding="utf-8")
    tmp.replace(OUT)
    print(f"wrote prayers.json for {date_str}")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception as exc:  # noqa: BLE001
        print(f"fatal: {exc}", file=sys.stderr)
        sys.exit(1)
